sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/FnApiStateAccessor.java

```diff
@@ -85,27 +83,64 @@ public FnApiStateAccessor(
       Map<TupleTag<?>, SideInputSpec> sideInputSpecMap,
       BeamFnStateClient beamFnStateClient,
       Coder<?> keyCoder,
-      Coder<BoundedWindow> windowCoder) {
+      Coder<BoundedWindow> windowCoder,
+      Supplier<WindowedValue<?>> currentElementSupplier,
+      Supplier<BoundedWindow> currentWindowSupplier) {
     this.pipelineOptions = pipelineOptions;
     this.stateKeyObjectCache = Maps.newHashMap();
     this.sideInputSpecMap = sideInputSpecMap;
     this.beamFnStateClient = beamFnStateClient;
     this.ptransformId = ptransformId;
     this.processBundleInstructionId = processBundleInstructionId;
     this.stateFinalizers = new ArrayList<>();
-
-    this.keyCoder = keyCoder;
-    this.windowCoder = windowCoder;
+    this.currentWindowSupplier = currentWindowSupplier;
+    this.encodedCurrentKeySupplier = memoizeFunction(currentElementSupplier,
+        element -> {
+          checkState(
+              element.getValue() instanceof KV,
+              "Accessing state in unkeyed context. Current element is not a KV: %s.",
+              element);
+          checkState(
+              keyCoder != null,
+              "Accessing state in unkeyed context, no key coder available");
+
+          ByteString.Output encodedKeyOut = ByteString.newOutput();
+          try {
+            ((Coder) keyCoder).encode(((KV<?, ?>) element.getValue()).getKey(), encodedKeyOut);
+          } catch (IOException e) {
+            throw new IllegalStateException(e);
+          }
+          return encodedKeyOut.toByteString();
+        });
+
+    this.encodedCurrentWindowSupplier = memoizeFunction(
+        currentWindowSupplier,
+        window -> {
+          ByteString.Output encodedWindowOut = ByteString.newOutput();
+          try {
+            windowCoder.encode(window, encodedWindowOut);
+          } catch (IOException e) {
+            throw new IllegalStateException(e);
+          }
+          return encodedWindowOut.toByteString();
+        });
   }
 
-  public void setCurrentElement(WindowedValue<?> currentElement) {
-    this.currentElement = currentElement;
-    this.encodedCurrentKey = null;
-  }
-
-  public void setCurrentWindow(BoundedWindow currentWindow) {
-    this.currentWindow = currentWindow;
-    this.encodedCurrentWindow = null;
+  private static <ArgT, ResultT> Supplier<ResultT> memoizeFunction(
```

Comment:
nit: Guava has memoize via Suppliers#memoize

And you can convert from Guava Supplier -> JDK suppler by referring to the `::get` method reference. Like shown in https://stackoverflow.com/questions/35331327/does-java-8-have-cached-support-for-suppliers

Reply:
I considered Suppliers.memoize but it doesn't do what I want - it memoizes the result of .get() (if you call get() again it will keep returning the same value), whereas I want sort of a one-key cache (if the argument is the same, then reuse, else recompute). I couldn't find anything standard for this.